### PR TITLE
Fix Pkg.build when run on a non-installed package

### LINKED ISSUE #15948: Pkg.build on non-existent package causes SIGTERM
On a fresh build of julia:
```
julia> Pkg.build("FooBar")

signal (15): Terminated
while loading no file, in expression starting on line 0
_ZN4llvm17PMTopLevelManager17findAnalysisUsageEPNS_4PassE at /root/julia/usr/bin/../lib/libLLVM-3.7.1.so (unknown line)
_ZN4llvm13PMDataManager23collectRequiredAnalysisERNS_15SmallVectorImplIPNS_4PassEEERNS1_IPKvEES3_ at /root/julia/usr/bin/../lib/libLLVM-3.7.1.so (unknown line)
_ZN4llvm13PMDataManager3addEPNS_4PassEb at /root/julia/usr/bin/../lib/libLLVM-3.7.1.so (unknown line)
_ZN4llvm17PMTopLevelManager12schedulePassEPNS_4PassE at /root/julia/usr/bin/../lib/libLLVM-3.7.1.so (unknown line)
_ZN4llvm17PMTopLevelManager12schedulePassEPNS_4PassE at /root/julia/usr/bin/../lib/libLLVM-3.7.1.so (unknown line)
_ZN4llvm16TargetPassConfig7addPassEPNS_4PassEbb at /root/julia/usr/bin/../lib/libLLVM-3.7.1.so (unknown line)
_ZN4llvm16TargetPassConfig7addPassEPKvbb at /root/julia/usr/bin/../lib/libLLVM-3.7.1.so (unknown line)
_ZN4llvm16TargetPassConfig26addMachineLateOptimizationEv at /root/julia/usr/bin/../lib/libLLVM-3.7.1.so (unknown line)
_ZN4llvm16TargetPassConfig16addMachinePassesEv at /root/julia/usr/bin/../lib/libLLVM-3.7.1.so (unknown line)
unknown function (ip: 0x7f38851cac53)
_ZN4llvm17LLVMTargetMachine17addPassesToEmitMCERNS_6legacy15PassManagerBaseERPNS_9MCContextERNS_17raw_pwrite_streamEb at /root/julia/usr/bin/../lib/libLLVM-3.7.1.so (unknown line)
JuliaOJIT at /root/julia/src/./jitlayers.cpp:336
jl_init_codegen at /root/julia/src/codegen.cpp:5766
_julia_init at /root/julia/src/init.c:623
julia_init at /root/julia/src/task.c:278
unknown function (ip: 0x402679)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x401708)
unknown function (ip: 0xffffffffffffffff)
Allocations: 73339 (Pool: 73238; Big: 101); GC: 0
ERROR: Base.Pkg.PkgError("FooBar is not an installed package")
 [inlined code] from ./pkg.jl:13
 in build!(::Array{ASCIIString,1}, ::Pipe, ::Set{Any}) at ./pkg/entry.jl:506
 in build!(::Array{ASCIIString,1}, ::Dict{Any,Any}, ::Set{Any}) at ./pkg/entry.jl:551
 [inlined code] from ./dict.jl:513
 in build(::Array{ASCIIString,1}) at ./pkg/entry.jl:573
 [inlined code] from ./promotion.jl:229
 in (::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#build,Tuple{Array{ASCIIString,1}}})() at ./pkg/dir.jl:31
 in cd(::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#build,Tuple{Array{ASCIIString,1}}}, ::ASCIIString) at ./file.jl:48
 in #cd#1(::Array{Any,1}, ::Any, ::Any, ::Array{ASCIIString,1}, ::Vararg{Array{ASCIIString,1}}) at ./pkg/dir.jl:31
 [inlined code] from ./pkg/dir.jl:25
 in build(::ASCIIString, ::Vararg{ASCIIString}) at ./pkg.jl:210
 in eval(::Module, ::Any) at ./boot.jl:236
```

### PR DESCRIPTION
**Pull request by omus**
Fixes #15948

### CODE REVIEW

**Comment by omus** [comment]
Should be safe to backport to 0.4

**Comment by tkelman** [comment]
Linux 32 failure looks real - `Expression: !(contains(msg,"signal (15)"))` ?

**Comment by omus** [comment]
New test cases caught their first failure. I've updated the code to attempt to gracefully exit from the detached process.

File: base/pkg/entry.jl [hunk 1]
```diff
@@ -582,8 +582,8 @@ function build!(pkgs::Vector, errs::Dict, seen::Set=Set())
             end
         end
     catch err
-        kill(pobj)
         close(io)
+        isa(err, PkgError) ? wait(pobj) : kill(pobj)
```
Comment (tkelman) [inline base/pkg/entry.jl]: I think since `PkgError` doesn't exist on release-0.4, backporting this won't quite work

Comment (omus) [inline base/pkg/entry.jl]: Would backporting `PkgError` be a reasonable fix?

Comment (omus) [inline base/pkg/entry.jl]: I introduced `PkgError` into 0.4 as the fix. See #16691

**Pull request merged**